### PR TITLE
Add initial implementation for EXTLINUX bootloader config

### PR DESCRIPTION
**Pull request by matwey**
Hello,

This is to add initial support for ExtLinux config file which is planned to be used with u-boot on openSUSE ARMv7.

### CODE REVIEW

**Comment by arvidjaar** [comment]
Sorry for off-topic question, but - any reason you do not use grub2 there?

File: src/Path.pm [hunk 1]
```diff
@@ -126,6 +126,18 @@ sub Sysconfig {
 }
 
 =item
+C<< $path = Bootloader::Path::Extlinux_conf(); >>
+
+Gets path to lilo(or powerlilo) configuration file.
```
Comment (StefanBruens) [inline src/Path.pm]: lilo?

**Comment by afaerber** [comment]
@arvidjaar: Wherever we have a UEFI bootloader available, we _are_ using grub2. This however is for U-Boot bootloader, which is stopping to support its by default disabled `CONFIG_API` option that grub2's `uboot` platform support relies on - I reported this on their mailing list and they refused to fix, e.g., sunxi for v2015.10, pointing to this extlinux.conf mechanism, which they [document](http://git.denx.de/?p=u-boot.git;a=blob;f=doc/README.distro;h=9e4722a86ee562fdc80b675f5be9d30e30e61597;hb=HEAD) as their preference for distros over the single-kernel boot.scr mechanism we use today. Another problem with grub2 is that it needs to be patched per board, so a single package does not work. See for example my [Jetson TK1 grub2 installation guide](https://en.opensuse.org/HCL:JetsonTK1#Installing_GRUB2). Even worse, my modified grub2 builds proved to be rather unreliable on the Jetson TK1, often not getting to any prompt or menu. If you want to discuss grub2 further, I suggest you post to opensuse-arm mailing list instead.

File: perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ONTIMEOUT localboot
+DEFAULT default
+MENU TITLE Boot menu
+PROMPT 0
+
+LABEL default
+         MENU LABEL Default
+         LINUX /boot/zImage
+         FDTDIR /boot/dtb/
+         FDT exynos5250-snow.dtb
```
Comment (StefanBruens) [inline perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf]: if FDT is specified, it is the whole path, fdtdir has no effect.
http://code.metager.de/source/xref/denx/u-boot/doc/README.pxe#158

fdtdir is especially useful for the RPi, the fdt filename is generated by u-boot from OTP (one-time-programmable) memory in the SoC.
http://git.denx.de/?p=u-boot.git;a=blob;f=board/raspberrypi/rpi/rpi.c#l77

In fact, fdtdir should be used in general.

Comment (afaerber) [inline perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf]: While I agree that FDT should not unnecessarily be specified, does it do any harm in the testsuite here? If I understood perl-bootloader correctly, it needs to parse and retain anything a user put into the file, so having a test case covering everything seems good.

Comment (StefanBruens) [inline perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf]: Maybe use fdt and fdtdir in independent sections? Or even issue a warning when parsing such a file, as it is not really well-formed?

File: src/Path.pm (hunk 1 above)
Comment (matwey) [inline src/Path.pm]: Nice catch, thank you.

File: perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf (hunk 2 above)
Comment (matwey) [inline perl-Bootloader-testsuite/tests/test_interface/backup/fake_root1/boot/extlinux/extlinux.conf]: Good point. There should be either FDTDIR or FDT in a section. I have to sort this out.

**Comment by wfeldt** [comment]
matwey, thanks for the patch.

Please note, though, that we are planning to discontinue perl-Bootloader in its current form. The plan is to substitute the perl library with a set of simple shell scripts. yast2 will no longer use perl-Bootloader at all and just work on the bootloader config filers directly (this with grub2 in mind).

A preliminary version is in the test_10 branch atm. I was about to release it into tuimbleweed in the hext days to get some attention. For the time being, it's fully backward compatible by keeping the perl code for non-grub2-llike loaders.

See https://github.com/openSUSE/perl-bootloader/blob/test_10/new_pbl.txt for a short description.

**Comment by matwey** [comment]
@wfeldt What do you think, would it be better if I implemented extlinux support using new approach?

**Comment by jreidinger** [comment]
@matwey yes, keep it as simple as possible. Also expend that without feature installer won't support another bootloader so it need manual configuration. Perfect case will be if extlinux itself have script to update to new kernel and just simple call it.

**Comment by matwey** [comment]
What is about LILO or GRUB? How are you planning to support bootloaders not able to find new kernel in a manner like GRUB2 does?

There is no such a script for ExtLinux, I thought that ```update-bootloader``` did the job.

**Comment by wfeldt** [comment]
matwey, yes. Although, as said, I'l lkeep the old code for now. But yast2 might stop working at some point when Josef gets around dropping the perl-Bootloader usage in yast2.

**Comment by wfeldt** [comment]
That's true. update-bootloader will continue to work for now. If time permits I'll look into adding a grub1 script.

**Comment by wfeldt** [comment]
Note this mainly affects yast2 bootloader, where no support for loaders != grub2 is planned.

**Comment by jreidinger** [comment]
@wfeldt @matwey in general it should be easier to allow all bootloaders to manage their entries at own instead of having single script that support all of them. In general for grub1 and lilo you can still
1) update it manually
2) use symlinks ( my favorite way )
3) use e.g. grubby to manage it

**Comment by wfeldt** [comment]
We might in the end find us doing grub2-like wrapper scripts for the other loaders...

**Comment by matwey** [comment]
However, what is about kernel RPM post/pre scripts? What interface they are supposed to use?

**Comment by wfeldt** [comment]
They use update-bootloader and continue to work.

**Comment by jsrain** [comment]
@matwey From kernel script's POV, nothing changes, but the internal implementation will be simplified (and only support GRUB2). It could be possible to create e.g. perl-Bootloader-extlinux package, which would contain implementation of the same scripts specific for extlinux bootloader.
Just note that YaST will be changed as well and will only work with GRUB2.

**Comment by matwey** [comment]
@wfeldt kernel scripts call ```bootloader_entry [add|remove] <kernel-flavor> <kernel-release> <image-name> <initrd-name>``` as far as I understand. I am right that new PBL just drops this arguments?

**Comment by wfeldt** [comment]
@matwey, yes. But that is just emulating the situation we have with grub2, because grub2 handles the config all by itself. I'm not at all against passing on those arguments to the scripts if needed.

As only grub2 is officially supported, there's nothing at all to say against doing it differently with other loaders (like grub1). Whatever logic is needed will end up in the per-loader scripts anyway.

That said, do you want me to merge your pull request? ATM there are no conflicts and it doesn't affect the other loaders. So this would be fine with me.

**Comment by matwey** [comment]
When will you plan to deploy new pbl to Factory?

**Comment by wfeldt** [comment]
I already did. It's currently in a staging project and will be in Factory soon.

**Comment by matwey** [comment]
Ok then. I've started to implement it like this:
https://github.com/matwey/perl-bootloader/commit/32493d6721a833c1ad5d46db18f48ea66011c3ea
So, I think we don't need this PR anymore.

**Comment by arvidjaar** [comment]
@afaerber 

>Another problem with grub2 is that it needs to be patched per board

beta3 generates position independent self relocatable image on arm-uboot

> I suggest you post to opensuse-arm mailing list instead

It does not like me :)

**Pull request closed without merging**